pkg/revisions/tag_watcher.go

```diff
@@ -0,0 +1,118 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package revisions
+
+import (
+	"sync"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/cache"
+
+	"istio.io/api/label"
+	"istio.io/istio/istioctl/pkg/tag"
+	"istio.io/istio/pkg/kube"
+	"istio.io/istio/pkg/kube/controllers"
+	"istio.io/pkg/log"
+)
+
+// TagWatcher keeps track of the current tags and can notify watchers
+// when the tags change.
+type TagWatcher interface {
+	Run(stopCh <-chan struct{})
+	HasSynced() bool
+	AddHandler(handler TagHandler)
+}
+
+// TagHandler is a callback for when the tags revision change.
+type TagHandler func([]string)
+
+type tagWatcher struct {
+	revision string
+	handlers []TagHandler
+
+	queue           controllers.Queue
+	webhookInformer cache.SharedIndexInformer
+	mu              sync.RWMutex
+	tagsToRevisions map[string]string
+	revisionsToTags map[string][]string
+}
+
+func NewTagWatcher(client kube.Client, revision string) TagWatcher {
+	p := &tagWatcher{
+		revision:        revision,
+		mu:              sync.RWMutex{},
+		tagsToRevisions: map[string]string{},
+		revisionsToTags: map[string][]string{},
+	}
+	p.queue = controllers.NewQueue("tag", controllers.WithReconciler(p.updateTags))
+	p.webhookInformer = client.KubeInformer().Admissionregistration().V1().MutatingWebhookConfigurations().Informer()
```

Comment:
nit: can use the new kclient

Reply:
Actually its not really a nit, HasSynced is currently incorrect without it

Reply:
In that case, we should update default_watcher, which is the source of this code...